Framework/DataHandling/src/SetSample.cpp

```diff
@@ -323,87 +371,42 @@ SetSample::createFlatPlateXML(const Kernel::PropertyManager &args) const {
 /**
  * Create the XML required to define a cylinder from the given args
  * @param args A user-supplied dict of args
+ * @param hollow True if an annulus is to be created
  * @return The XML definition string
  */
 std::string
-SetSample::createCylinderXML(const Kernel::PropertyManager &args) const {
-  double height = args.getProperty("Height");
-  double radius = args.getProperty("Radius");
-  std::vector<double> center = args.getProperty("Center");
-  // Expected to be a long so that the mapping from Python is simple
-  // Expected to be a long so that the mapping from Python is simple
-  long axisIdx = getAxisIndex(args);
-  // convert to metres
-  height *= 0.01;
-  radius *= 0.01;
-  std::transform(center.begin(), center.end(), center.begin(),
-                 [](double val) { return val *= 0.01; });
-
-  // Shift so that cylinder is centered at center position
-  const double cylinderBase = (-1e-03 * height) + center[axisIdx];
-
-  std::ostringstream xmlShapeStream;
-  xmlShapeStream << "<cylinder id=\"sample-shape\"> "
-                 << "<centre-of-bottom-base x=\"" << center[axisIdx]
-                 << "\" y=\"" << cylinderBase << "\" z=\"" << center[axisIdx]
-                 << "\" /> "
-                 << "<axis ";
-
-  if (axisIdx == 0)
-    xmlShapeStream << "x=\"1\" y=\"0\" z=\"0\" /> ";
-  else if (axisIdx == 1)
-    xmlShapeStream << "x=\"0\" y=\"1\" z=\"0\" /> ";
-  else
-    xmlShapeStream << "x=\"0\" y=\"0\" z=\"1\" /> ";
-
-  xmlShapeStream << "<radius val=\"" << radius << "\" /> "
-                 << "<height val=\"" << height << "\" /> "
-                 << "</cylinder>";
-
-  return xmlShapeStream.str();
-}
-
-/**
- * Create the XML required to define an annulus from the given args
- * @param args A user-supplied dict of args
- * @return The XML definition string
- */
-std::string
-SetSample::createHollowCylinderXML(const Kernel::PropertyManager &args) const {
+SetSample::createCylinderLikeXML(const Kernel::PropertyManager &args,
+                                 bool hollow) const {
+  const std::string tag = hollow ? "hollow-cylinder" : "cylinder";
   double height = args.getProperty("Height");
-  double innerRadius = args.getProperty("InnerRadius");
-  double outerRadius = args.getProperty("OuterRadius");
-  std::vector<double> center = args.getProperty("Center");
-  // Expected to be a long so that the mapping from Python is simple
+  double innerRadius = hollow ? args.getProperty("InnerRadius") : 0.0;
+  double outerRadius =
+      hollow ? args.getProperty("OuterRadius") : args.getProperty("Radius");
+  std::vector<double> centre = args.getProperty("Center");
   long axisIdx = getAxisIndex(args);
```

Comment:
@quantumsteve No it's not ideal. It's a `long` because the mapping from Python creates a `long` from a Python integer and this avoids then having to do much work checking and converting things. I see the problem as the inflexibility of the `PropertyWithValue` classes really.